### PR TITLE
add Fibonacci test for actions

### PR DESCRIPTION
**Pull request by wjwwood**
Requires ros2/rclcpp#593
Requires ros2/rclcpp#594

### CODE REVIEW

**Comment by wjwwood** [comment]
This is ready for review.

File: test_communication/test/test_action_client.cpp [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+  std::vector<int32_t> valid_fibo_seq;
+  valid_fibo_seq.push_back(0);
+  valid_fibo_seq.push_back(1);
+  for (size_t i = 1; i < order; ++i) {
+    valid_fibo_seq.push_back(valid_fibo_seq[i] + valid_fibo_seq[i - 1]);
+  }
+
+  {
+    ActionClientTest<test_msgs::action::Fibonacci> test;
+    size_t order = 10;
```
Comment (hidmic) [inline test_communication/test/test_action_client.cpp]: @wjwwood hiding above's variable?

File: test_communication/test/test_action_client.cpp [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+            fprintf(
+              stderr,
+              "feedback sequence not correct, expected %d but got %d for order %zu\n",
+              valid_fibo_seq[i], feedback->sequence[i], i);
+            return false;
+          }
+        }
+        return true;
+      };
+    result.push_back(test);
```
Comment (hidmic) [inline test_communication/test/test_action_client.cpp]: @wjwwood only one test?

File: test_communication/test/test_action_client.cpp (hunk 1 above)
Comment (wjwwood) [reply test_communication/test/test_action_client.cpp]: Yeah, that was intentional, the idea being each scoped section would make their own order, but I guess I could use different names here.

File: test_communication/test/test_action_client.cpp (hunk 2 above)
Comment (wjwwood) [reply test_communication/test/test_action_client.cpp]: Yes.

**Comment by wjwwood** [comment]
> Will the test generators output more than one element eventually?

That's the plan. I based this on the tests from services, which have a "add two int" test and a more complete one that tests all IDL types in a single service call. I didn't do that yet for actions because of the time and because it's somewhat redundant given that the tests for topics and services already test a complex message like that and actions derive from services and topics.

File: test_communication/test/test_action_server.cpp [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+      RCLCPP_ERROR(logger, "unexpected goal request");
+      should_fail = true;
+      return rclcpp_action::GoalResponse::REJECT;
+    };
+
+  auto cancel_callback =
+    [logger](auto /* goal_handle */) {
+      RCLCPP_ERROR(logger, "received unexpected request to cancel goal");
+      return rclcpp_action::CancelResponse::ACCEPT;
+    };
```
Comment (sloretz) [inline test_communication/test/test_action_server.cpp]: `should_fail = true` if cancel request received?

File: test_communication/CMakeLists.txt [hunk 4]
```diff
@@ -226,6 +231,45 @@ if(BUILD_TESTING)
...
+        "${CMAKE_CURRENT_BINARY_DIR}/test_action_client_server${test_suffix}_$<CONFIG>.py"
+        PYTHON_EXECUTABLE "${_PYTHON_EXECUTABLE}"
+        APPEND_LIBRARY_DIRS "${append_library_dirs}"
+        TIMEOUT 30
+        ${SKIP_TEST})
+      if(TEST test_action_client_server${test_suffix})
+        set_tests_properties(
+          test_action_client_server${test_suffix}
+          PROPERTIES DEPENDS
+            "test_requester_cpp__${rmw_implementation1};test_replier_cpp__${rmw_implementation2}"
```
[the remark this answers is not in the document]
Comment (sloretz) [reply test_communication/CMakeLists.txt]: 7c424cd

[the remark this answers is not in the document]
Comment (wjwwood) [reply test_communication/CMakeLists.txt]: lgtm thanks!

File: test_communication/test/test_action_client.cpp [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+    RCLCPP_INFO(logger, "sent goals for %f seconds\n", diff.count());
+  });
+
+  while (rclcpp::ok() && test_index < goal_tests.size()) {
+    RCLCPP_INFO(logger, "sending goal #%zu", test_index + 1);
+
+    // on feedback, check the feedback is valid
+    auto feedback_callback =
+      [&](auto, const auto & feedback) {
+        RCLCPP_INFO(logger, "received feedback");
```
[the remark this answers is not in the document]
Comment (wjwwood) [reply test_communication/test/test_action_client.cpp]: Yeah, I wasn't asserting that at all. I just wanted to make sure that if feedback is sent it is valid. I could keep a counter, but I'm not sure if all the feedback would be received before the result or not, and if not how I could wait for all the feedback even after the result.

I'd consider this a good improvement on the test if we could find a good way to do it.

File: test_communication/test/test_action_server.cpp (hunk 3 above)
Comment (wjwwood) [reply test_communication/test/test_action_server.cpp]: Probably.

**Comment by hidmic** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/5818/)
* Linux-aarch64 [Build Status](https://ci.ros2.org/job/ci_linux-aarch64/2384/)
* macOS **Waiting...**
* Windows **Waiting...**

**Comment by sloretz** [comment]
CI with fix for test failure when using connext

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/5828/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/2393/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/4820/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/5723/)

**Comment by sloretz** [comment]
CI post ros2/rosidl#338
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/5833/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/2397/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/4824/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/5727/)

**Comment by sloretz** [comment]
Rebuilding just windows to check if warning is gone [Build Status](https://ci.ros2.org/job/ci_windows/5729/)

**Pull request merged**